### PR TITLE
matrix: send the display name instead of the user name

### PR DESCRIPTION
**Pull request by nightmared**
This PR set the default behavior to sending the matrix display name instead of the user name, as asked by #441.

For example, this means '@thobys:server' will be translated to the nickname I chose on `server`.

In addition, if the user didn't set a display name, this will fallback to the user name.

To turn off this behavior, we use the option `UseUserName`, as it seems to be the very meaning this option was intended for in the discord bridge.

### CODE REVIEW

File: bridge/matrix/matrix.go [hunk 1]
```diff
@@ -216,6 +225,73 @@ func (b *Bmatrix) handlematrix() {
...
+	b.RLock()
+	for k, v := range b.NicknameMap {
+		if now.Sub(v.lastUpdated) > 10*time.Minute {
+			toDelete = append(toDelete, k)
+		}
+	}
+	b.RUnlock()
+
+	b.Lock()
+	if len(toDelete) != 0 {
```
Comment (42wim) [inline bridge/matrix/matrix.go:242]: you can remove this if, if the len is 0 the for loop will not run

File: bridge/matrix/matrix.go [hunk 2]
```diff
@@ -216,6 +225,73 @@ func (b *Bmatrix) handlematrix() {
 	}()
 }
 
+func (b *Bmatrix) setDisplayName(mxid string, displayName string) string {
```
Comment (42wim) [inline bridge/matrix/matrix.go:228]: maybe cacheDisplayName is better as function name (just a suggestion)

File: bridge/matrix/matrix.go [hunk 3]
```diff
@@ -216,6 +225,73 @@ func (b *Bmatrix) handlematrix() {
...
+}
+
+func (b *Bmatrix) getDisplayName(mxid string) string {
+	if b.GetBool("UseUserName") {
+		return mxid[1:]
+	}
+
+	b.RLock()
+	if val, present := b.NicknameMap[mxid]; present {
+		if time.Since(val.lastUpdated) < 10*time.Minute {
```
Comment (42wim) [inline bridge/matrix/matrix.go:263]: We don't need the time check here as you're already cleaning up outdated nicks, you can remove this if?

File: bridge/matrix/matrix.go [hunk 4]
```diff
@@ -216,6 +225,73 @@ func (b *Bmatrix) handlematrix() {
...
+}
+
+func (b *Bmatrix) handleMemberChange(ev *matrix.Event) {
+	// Update the displayname on join messages, according to https://matrix.org/docs/spec/client_server/r0.6.1#events-on-change-of-profile-information
+	if ev.Content["membership"] == "join" {
+		dn, dnPresent := ev.Content["displayname"]
+		dnStr, dnValidString := dn.(string)
+		if dnPresent && dnValidString {
+			b.setDisplayName(ev.Sender, dnStr)
+		}
```
Comment (42wim) [inline bridge/matrix/matrix.go:291]: not tested, but shouldn't this work also?

```golang
if dn,ok := ev.Content["displayname"].(string);ok {
    b.setDisplayName(ev.Sender, dn)
}
```

**Review by 42wim — CHANGES_REQUESTED** [review]

**Comment by 42wim** [comment]
@nightmared thanks for you work! 👍 
I've got a few comments

File: bridge/matrix/matrix.go (hunk 4 above)
Comment (nightmared) [reply bridge/matrix/matrix.go:291]: Oh, I didn't know one could chain the presence check and the cast check like, that's nice ;)

**Comment by 42wim** [comment]
looks good, can you rebase as I just merged your previous PR ? :)

**Review by 42wim — APPROVED** [review]

**Comment by 42wim** [comment]
Could you also add the `UseUsername` option to the discord section in matterbridge.toml.sample ?

**Comment by nightmared** [comment]
Hmm, maybe I should move some functions to a new `helpers.go` to make `codeclimate` happy.

**Comment by 42wim** [comment]
LGTM, thank you.
As it is a breaking change in behavior I'm going to put it in release 1.20.0 instead of 1.19.1
I'll merge it after 1.19.1 is released.

**Pull request merged**